finite_difference/src/grid_mod.f90

```diff
@@ -244,6 +255,26 @@ function grid_constructor(grid_name, &
     ! subject to.
     self%boundary_conditions(1:3) = boundary_conditions(1:3)
 
+    ! Ensure pointers to device memory are zeroed so we know whether or
+    ! not the buffers have been created
+    self%dx_t_device = 0
+    self%dy_t_device = 0
+    self%dx_u_device = 0
+    self%dy_u_device = 0
+    self%dx_v_device = 0
+    self%dy_v_device = 0
+    self%dx_f_device = 0
+    self%dy_f_device = 0
+    self%area_t_device = 0
+    self%area_u_device = 0
+    self%area_v_device = 0
+    self%gphiu_device = 0
+    self%gphiv_device = 0
+    self%gphif_device = 0
+    self%xt_device = 0
+    self%yt_device = 0
+    self%tmask_device = 0
```

Comment:
Maybe the C binding `c_null_ptr` is more appropriate? and shouldn't the `grid_init` function allocate and initialize these values? (I am not sure when are they allocated at the moment)

Reply:
Oh yes, that's a good idea. However, gfortran then refused to compile the code from which I concluded that these device pointers should be declared as `type(c_ptr)`. If I change that then all of the PSyclone-generated code will probably have to be changed too. Therefore, I suggest we leave it as is for the moment.

Reply:
To answer your second question, the initialisation of these pointers happens when required in the code generated by PSyclone - we only do the OpenCL device setup the first time an Invoke is called so we don't have a device until then.